In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


#### Importing the Libaries I will need off the bat

# Importing the Data

In [3]:
df = pd.read_csv("valorant_data.csv")


C:\Users\Connor\AppData\Local\Temp\ipykernel_8492\2196491408.py:1: DtypeWarning: Columns (7,17,18) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("valorant_data.csv")


#### Link to valorant dataset 
#### https://www.kaggle.com/datasets/qualidea1217/valorant-pro-matches-since-april-2021

In [4]:
num_instances = df.shape[0]
print("Number of instances:", num_instances)

Number of instances: 249710


#### Determining the number of instances in the dataset

In [5]:
df.head()

Unnamed: 0   match-datetime  patch    map       team1        team2  \
0           0  4/16/2023 10:00   6.06  Haven  Impulse GW  EGN Esports   
1           1  4/16/2023 10:00   6.06  Haven  Impulse GW  EGN Esports   
2           2  4/16/2023 10:00   6.06  Haven  Impulse GW  EGN Esports   
3           3  4/16/2023 10:00   6.06  Haven  Impulse GW  EGN Esports   
4           4  4/16/2023 10:00   6.06  Haven  Impulse GW  EGN Esports   

   team1-score team2-score  Team 1 Win  Team 2 Win  ... hs-ct first kill fk-t  \
0           13           6           1           0  ...  0.26        1.0  0.0   
1           13           6           1           0  ...  0.38        0.0  0.0   
2           13           6           1           0  ...  0.27        2.0  2.0   
3           13           6           1           0  ...  0.24        7.0  3.0   
4           13           6           1           0  ...  0.80        0.0  0.0   

   fk-ct  first death  fd-t  fd-ct first kills minus first deaths fkmd-t  \
0    1.0          1.0   0.0    1.0                            0.0    0.0   
1    0.0          1.0   1.0    0.0                           -1.0   -1.0   
2    0.0          1.0   1.0    0.0                            1.0    1.0   
3    4.0          5.0   5.0    0.0                            2.0   -2.0   
4    0.0          1.0   0.0    1.0                           -1.0    0.0   

   fkmd-ct  
0      0.0  
1      0.0  
2      0.0  
3      4.0  
4     -1.0  

[5 rows x 49 columns]

### Displaying the first 5 rows of the dataset

In [6]:
df.tail()

Unnamed: 0  match-datetime  patch   map      team1       team2  \
249705      249705  4/14/2021 5:30   2.07  Bind  Mindfreak  Team Bliss   
249706      249706  4/14/2021 5:30   2.07  Bind  Mindfreak  Team Bliss   
249707      249707  4/14/2021 5:30   2.07  Bind  Mindfreak  Team Bliss   
249708      249708  4/14/2021 5:30   2.07  Bind  Mindfreak  Team Bliss   
249709      249709  4/14/2021 5:30   2.07  Bind  Mindfreak  Team Bliss   

        team1-score team2-score  Team 1 Win  Team 2 Win  ... hs-ct first kill  \
249705           14          12           1           0  ...  0.38        2.0   
249706           14          12           1           0  ...  0.21        4.0   
249707           14          12           1           0  ...  0.13        4.0   
249708           14          12           1           0  ...  0.44        2.0   
249709           14          12           1           0  ...  0.18        0.0   

       fk-t  fk-ct  first death  fd-t  fd-ct first kills minus first deaths  \
249705  2.0    0.0          2.0   2.0    0.0                            0.0   
249706  3.0    1.0          5.0   1.0    4.0                           -1.0   
249707  1.0    3.0          3.0   1.0    2.0                            1.0   
249708  0.0    2.0          3.0   2.0    1.0                           -1.0   
249709  0.0    0.0          1.0   1.0    0.0                           -1.0   

       fkmd-t  fkmd-ct  
249705    0.0      0.0  
249706    2.0     -3.0  
249707    0.0      1.0  
249708   -2.0      1.0  
249709   -1.0      0.0  

[5 rows x 49 columns]

### Displaying the last 5 rows of the dataset

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 249710 entries, 0 to 249709
Data columns (total 49 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Unnamed: 0                         249710 non-null  int64  
 1   match-datetime                     249710 non-null  object 
 2   patch                              249710 non-null  float64
 3   map                                249710 non-null  object 
 4   team1                              249710 non-null  object 
 5   team2                              249710 non-null  object 
 6   team1-score                        249710 non-null  int64  
 7   team2-score                        249710 non-null  object 
 8   Team 1 Win                         249710 non-null  int64  
 9   Team 2 Win                         249710 non-null  int64  
 10  player-name                        249710 non-null  object 
 11  player-team                        2497

### Displaying the information of the dataset

# Cleaning the data for use in my models

In [8]:
df.isnull().sum()

Unnamed: 0                              0
match-datetime                          0
patch                                   0
map                                     0
team1                                   0
team2                                   0
team1-score                             0
team2-score                             0
Team 1 Win                              0
Team 2 Win                              0
player-name                             0
player-team                             0
agent                                   0
rating                                450
rating-t                              460
rating-ct                             460
Average combat score                  282
acs-t                                 351
acs-ct                               1158
kills                                  30
k-t                                    30
k-ct                                   30
deaths                                 30
d-t                               

### Checking for missing values in the dataset

In [9]:
(df.isnull().sum()/len(df))*100

Unnamed: 0                           0.000000
match-datetime                       0.000000
patch                                0.000000
map                                  0.000000
team1                                0.000000
team2                                0.000000
team1-score                          0.000000
team2-score                          0.000000
Team 1 Win                           0.000000
Team 2 Win                           0.000000
player-name                          0.000000
player-team                          0.000000
agent                                0.000000
rating                               0.180209
rating-t                             0.184214
rating-ct                            0.184214
Average combat score                 0.112931
acs-t                                0.140563
acs-ct                               0.463738
kills                                0.012014
k-t                                  0.012014
k-ct                              

### Checking for the percentage of missing values in the dataset

In [10]:
df.describe()

Unnamed: 0          patch    team1-score     Team 1 Win  \
count  249710.000000  249710.000000  249710.000000  249710.000000   
mean   124854.500000       4.028083      10.931080       0.580433   
std     72085.212197       1.171391       3.439841       0.493489   
min         0.000000       2.050000       0.000000       0.000000   
25%     62427.250000       3.060000       9.000000       0.000000   
50%    124854.500000       4.040000      13.000000       1.000000   
75%    187281.750000       5.060000      13.000000       1.000000   
max    249709.000000       6.060000      23.000000       1.000000   

          Team 2 Win         rating       rating-t      rating-ct  \
count  249710.000000  249260.000000  249250.000000  249250.000000   
mean        0.419326       1.004783       1.046198       1.249597   
std         0.493450       1.338180       4.297961       7.118967   
min         0.000000       0.010000       0.000000       0.000000   
25%         0.000000       0.780000       0.660000       0.740000   
50%         0.000000       0.980000       0.930000       1.010000   
75%         1.000000       1.200000       1.220000       1.310000   
max         1.000000     289.000000     319.000000     564.000000   

       Average combat score          kills  ...          hs-ct     first kill  \
count         249428.000000  249680.000000  ...  247989.000000  249380.000000   
mean             201.859575      14.575184  ...       0.255839       2.062539   
std               62.234810       5.522574  ...       0.151182       1.777457   
min                0.000000       0.000000  ...       0.000000       0.000000   
25%              159.000000      11.000000  ...       0.170000       1.000000   
50%              197.000000      14.000000  ...       0.240000       2.000000   
75%              240.000000      18.000000  ...       0.330000       3.000000   
max              691.000000      56.000000  ...       9.000000      17.000000   

                fk-t          fk-ct    first death           fd-t  \
count  249380.000000  249380.000000  249390.000000  249390.000000   
mean        0.938556       1.126169       2.062312       1.124780   
std         1.124959       1.198011       1.648656       1.204852   
min         0.000000       0.000000      -3.000000      -6.000000   
25%         0.000000       0.000000       1.000000       0.000000   
50%         1.000000       1.000000       2.000000       1.000000   
75%         1.000000       2.000000       3.000000       2.000000   
max        10.000000      11.000000      15.000000      10.000000   

               fd-ct  first kills minus first deaths         fkmd-t  \
count  249390.000000                   249380.000000  249265.000000   
mean        0.936918                       -0.000040      -0.187592   
std         1.018626                        2.149748       1.501666   
min        -6.000000                      -11.000000      -9.000000   
25%         0.000000                       -1.000000      -1.000000   
50%         1.000000                        0.000000       0.000000   
75%         1.000000                        1.000000       1.000000   
max         9.000000                       14.000000      10.000000   

             fkmd-ct  
count  249125.000000  
mean        0.187878  
std         1.519661  
min        -9.000000  
25%        -1.000000  
50%         0.000000  
75%         1.000000  
max        11.000000  

[8 rows x 39 columns]

### Displaying the summary statistics of the dataset

In [11]:
df = df.drop(df.columns[:6], axis=1)

### Drop first 6 columns in the dataset

In [12]:
df.head()

team1-score team2-score  Team 1 Win  Team 2 Win player-name player-team  \
0           13           6           1           0   HolyM0Ly   Impulse GW   
1           13           6           1           0      Dolfo   Impulse GW   
2           13           6           1           0    zePiCzz   Impulse GW   
3           13           6           1           0     s2newb   Impulse GW   
4           13           6           1           0    Minikid   Impulse GW   

     agent  rating  rating-t  rating-ct  ...  hs-ct first kill fk-t  fk-ct  \
0  Killjoy    1.32      1.21       1.51  ...   0.26        1.0  0.0    1.0   
1   Breach    1.22      1.21       1.24  ...   0.38        0.0  0.0    0.0   
2    Astra    1.13      1.14       1.13  ...   0.27        2.0  2.0    0.0   
3     Neon    1.06      0.35       2.29  ...   0.24        7.0  3.0    4.0   
4     Sova    0.81      0.51       1.31  ...   0.80        0.0  0.0    0.0   

   first death  fd-t  fd-ct  first kills minus first deaths  fkmd-t  fkmd-ct  
0          1.0   0.0    1.0                             0.0     0.0      0.0  
1          1.0   1.0    0.0                            -1.0    -1.0      0.0  
2          1.0   1.0    0.0                             1.0     1.0      0.0  
3          5.0   5.0    0.0                             2.0    -2.0      4.0  
4          1.0   0.0    1.0                            -1.0     0.0     -1.0  

[5 rows x 43 columns]

### Displaying the first 5 rows of the dataset after dropping the first 6 columns

In [13]:
df = df.drop(["player-name", "player-team", "agent"], axis=1)

### Drop columns "player-name", "player-team", "agent"

In [14]:
df.head()

team1-score team2-score  Team 1 Win  Team 2 Win  rating  rating-t  \
0           13           6           1           0    1.32      1.21   
1           13           6           1           0    1.22      1.21   
2           13           6           1           0    1.13      1.14   
3           13           6           1           0    1.06      0.35   
4           13           6           1           0    0.81      0.51   

   rating-ct  Average combat score  acs-t acs-ct  ...  hs-ct  first kill  \
0       1.51                 267.0  239.0  316.0  ...   0.26         1.0   
1       1.24                 239.0  256.0  212.0  ...   0.38         0.0   
2       1.13                 202.0  243.0  133.0  ...   0.27         2.0   
3       2.29                 267.0  174.0  428.0  ...   0.24         7.0   
4       1.31                 105.0   95.0  124.0  ...   0.80         0.0   

   fk-t  fk-ct  first death  fd-t  fd-ct  first kills minus first deaths  \
0   0.0    1.0          1.0   0.0    1.0                             0.0   
1   0.0    0.0          1.0   1.0    0.0                            -1.0   
2   2.0    0.0          1.0   1.0    0.0                             1.0   
3   3.0    4.0          5.0   5.0    0.0                             2.0   
4   0.0    0.0          1.0   0.0    1.0                            -1.0   

   fkmd-t  fkmd-ct  
0     0.0      0.0  
1    -1.0      0.0  
2     1.0      0.0  
3    -2.0      4.0  
4     0.0     -1.0  

[5 rows x 40 columns]

### Again checking the first 5 rows of the dataset after dropping the columns

In [15]:
df = df.drop(df.columns[-3:], axis=1)

### Drop last 3 columns in the dataset

In [16]:
df.head()

team1-score team2-score  Team 1 Win  Team 2 Win  rating  rating-t  \
0           13           6           1           0    1.32      1.21   
1           13           6           1           0    1.22      1.21   
2           13           6           1           0    1.13      1.14   
3           13           6           1           0    1.06      0.35   
4           13           6           1           0    0.81      0.51   

   rating-ct  Average combat score  acs-t acs-ct  ...  adr-ct  \
0       1.51                 267.0  239.0  316.0  ...   193.0   
1       1.24                 239.0  256.0  212.0  ...   115.0   
2       1.13                 202.0  243.0  133.0  ...   114.0   
3       2.29                 267.0  174.0  428.0  ...   244.0   
4       1.31                 105.0   95.0  124.0  ...    77.0   

   headshot percentage  hs-t  hs-ct  first kill  fk-t  fk-ct  first death  \
0                 0.31  0.35   0.26         1.0   0.0    1.0          1.0   
1                 0.41  0.43   0.38         0.0   0.0    0.0          1.0   
2                 0.33  0.38   0.27         2.0   2.0    0.0          1.0   
3                 0.27  0.32   0.24         7.0   3.0    4.0          5.0   
4                 0.39  0.23   0.80         0.0   0.0    0.0          1.0   

   fd-t  fd-ct  
0   0.0    1.0  
1   1.0    0.0  
2   1.0    0.0  
3   5.0    0.0  
4   0.0    1.0  

[5 rows x 37 columns]

### Checking the first 5 rows of the dataset after dropping the last 3 columns

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 249710 entries, 0 to 249709
Data columns (total 37 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   team1-score                        249710 non-null  int64  
 1   team2-score                        249710 non-null  object 
 2   Team 1 Win                         249710 non-null  int64  
 3   Team 2 Win                         249710 non-null  int64  
 4   rating                             249260 non-null  float64
 5   rating-t                           249250 non-null  float64
 6   rating-ct                          249250 non-null  float64
 7   Average combat score               249428 non-null  float64
 8   acs-t                              249359 non-null  object 
 9   acs-ct                             248552 non-null  object 
 10  kills                              249680 non-null  float64
 11  k-t                                2496

### Checking the information of the dataset after dropping some columns to figure out exactly what we are working with

In [18]:
df.isnull().sum()

team1-score                             0
team2-score                             0
Team 1 Win                              0
Team 2 Win                              0
rating                                450
rating-t                              460
rating-ct                             460
Average combat score                  282
acs-t                                 351
acs-ct                               1158
kills                                  30
k-t                                    30
k-ct                                   30
deaths                                 30
d-t                                   320
d-ct                                  320
assists                               320
a-t                                   320
a-ct                                  320
total kills minus deaths              321
tkmd-t                                579
tkmd-ct                               534
kills assists survive trade perct     357
kast-t                            

### Checking for missing values in the dataset

In [19]:
df.isna().sum()/len(df)*100

team1-score                          0.000000
team2-score                          0.000000
Team 1 Win                           0.000000
Team 2 Win                           0.000000
rating                               0.180209
rating-t                             0.184214
rating-ct                            0.184214
Average combat score                 0.112931
acs-t                                0.140563
acs-ct                               0.463738
kills                                0.012014
k-t                                  0.012014
k-ct                                 0.012014
deaths                               0.012014
d-t                                  0.128149
d-ct                                 0.128149
assists                              0.128149
a-t                                  0.128149
a-ct                                 0.128149
total kills minus deaths             0.128549
tkmd-t                               0.231869
tkmd-ct                           

### Checking for the percentage of missing values in the dataset

In [20]:
df = df.drop(["acs-ct", "acs-t", "kast-ct", "kast-t", "hs-t", "hs-ct", "adr-t","adr-ct"], axis=1)

### Dropping the columns with missing values over the threshold of 45% so imputing is easier

In [21]:
df.isna().sum()/len(df)*100

team1-score                          0.000000
team2-score                          0.000000
Team 1 Win                           0.000000
Team 2 Win                           0.000000
rating                               0.180209
rating-t                             0.184214
rating-ct                            0.184214
Average combat score                 0.112931
kills                                0.012014
k-t                                  0.012014
k-ct                                 0.012014
deaths                               0.012014
d-t                                  0.128149
d-ct                                 0.128149
assists                              0.128149
a-t                                  0.128149
a-ct                                 0.128149
total kills minus deaths             0.128549
tkmd-t                               0.231869
tkmd-ct                              0.213848
kills assists survive trade perct    0.142966
average damage per round          

### Checking for the percentage of missing values in the dataset

In [22]:
df['rating'] = df['rating'].fillna(df['rating'].mean())
df['rating-t'] = df['rating-t'].fillna(df['rating-t'].mean())
df['rating-ct'] = df['rating-ct'].fillna(df['rating-ct'].mean())
df['Average combat score'] = df['Average combat score'].fillna(df['Average combat score'].mean())
df['kills'] = df['kills'].fillna(df['kills'].mean())
df['k-t'] = df['k-t'].fillna(df['k-t'].mean())
df['k-ct'] = df['k-ct'].fillna(df['k-ct'].mean())
df['deaths'] = df['deaths'].fillna(df['deaths'].mean())
df['d-t'] = df['d-t'].fillna(df['d-t'].mean())
df['d-ct'] = df['d-ct'].fillna(df['d-ct'].mean())
df['assists'] = df['assists'].fillna(df['assists'].mean())
df['a-t'] = df['a-t'].fillna(df['a-t'].mean())
df['a-ct'] = df['a-ct'].fillna(df['a-ct'].mean())
df['total kills minus deaths'] = df['total kills minus deaths'].fillna(df['total kills minus deaths'].mean())
df['tkmd-t'] = df['tkmd-t'].fillna(df['tkmd-t'].mean())
df['tkmd-ct'] = df['tkmd-ct'].fillna(df['tkmd-ct'].mean())
df['kills assists survive trade perct'] = df['kills assists survive trade perct'].fillna(df['kills assists survive trade perct'].mean())
df['average damage per round'] = df['average damage per round'].fillna(df['average damage per round'].mean())
df['headshot percentage'] = df['headshot percentage'].fillna(df['headshot percentage'].mean())
df['first kill'] = df['first kill'].fillna(df['first kill'].mean())
df['fk-t'] = df['fk-t'].fillna(df['fk-t'].mean())
df['fk-ct'] = df['fk-ct'].fillna(df['fk-ct'].mean())
df['first death'] = df['first death'].fillna(df['first death'].mean())
df['fd-t'] = df['fd-t'].fillna(df['fd-t'].mean())
df['fd-ct'] = df['fd-ct'].fillna(df['fd-ct'].mean())

### Filled all missing values with the mean of the column

In [23]:
df = df.drop(["team1-score", "team2-score"], axis=1)

### Drop team1-score and team2-score

In [24]:
df.isna().sum()/len(df)*100

Team 1 Win                           0.0
Team 2 Win                           0.0
rating                               0.0
rating-t                             0.0
rating-ct                            0.0
Average combat score                 0.0
kills                                0.0
k-t                                  0.0
k-ct                                 0.0
deaths                               0.0
d-t                                  0.0
d-ct                                 0.0
assists                              0.0
a-t                                  0.0
a-ct                                 0.0
total kills minus deaths             0.0
tkmd-t                               0.0
tkmd-ct                              0.0
kills assists survive trade perct    0.0
average damage per round             0.0
headshot percentage                  0.0
first kill                           0.0
fk-t                                 0.0
fk-ct                                0.0
first death     

### Checking for the percentage of missing values in the dataset

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 249710 entries, 0 to 249709
Data columns (total 27 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Team 1 Win                         249710 non-null  int64  
 1   Team 2 Win                         249710 non-null  int64  
 2   rating                             249710 non-null  float64
 3   rating-t                           249710 non-null  float64
 4   rating-ct                          249710 non-null  float64
 5   Average combat score               249710 non-null  float64
 6   kills                              249710 non-null  float64
 7   k-t                                249710 non-null  float64
 8   k-ct                               249710 non-null  float64
 9   deaths                             249710 non-null  float64
 10  d-t                                249710 non-null  float64
 11  d-ct                               2497

### Checking the information of the dataset after filling missing values

In [26]:
df.to_csv('C:\\Users\\Connor\\Desktop\\valorant_data_cleaned.csv', index=False)

## Saving data so I do not need to continously rerun code

# Starting some predictions useing models 

In [27]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.tree import plot_tree
from sklearn.svm import SVR

### Importing the Libaries I will possibly need for making my models

In [28]:
X = df[['rating', 'rating-t', 'rating-ct', 'Average combat score', 'kills', 'k-t', 'k-ct', 'deaths',
        'd-t', 'd-ct', 'assists', 'a-t', 'a-ct', 'total kills minus deaths', 'tkmd-t', 'tkmd-ct', 'kills assists survive trade perct', 'average damage per round', 'headshot percentage', 'first kill', 'fk-t', 'fk-ct', 'first death', 'fd-t', 'fd-ct']]

### Making X copnsisting of columns: 'rating', 'rating-t', 'rating-ct', 'Average combat score', 'kills', 'k-t', 'k-ct', 'deaths', 'd-t', 'd-ct', 'assists', 'a-t', 'a-ct', 'total kills minus deaths', 'tkmd-t',
### 'tkmd-ct', 'kills assists survive trade perct', 'average damage per round', 'headshot percentage', 'first kill', 'fk-t', 'fk-ct', 'first death', 'fd-t', 'fd-ct'

In [29]:
y = df['Team 1 Win']

###  Split the data into training and testing sets

In [30]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)
regressor=DecisionTreeRegressor(random_state=42)
regressor.fit(X_train,y_train)

DecisionTreeRegressor(random_state=42)

### Training the model with the training data

In [31]:
y_pred=regressor.predict(X_test)

### Predict the model

In [32]:
df.head()

Team 1 Win  Team 2 Win  rating  rating-t  rating-ct  Average combat score  \
0           1           0    1.32      1.21       1.51                 267.0   
1           1           0    1.22      1.21       1.24                 239.0   
2           1           0    1.13      1.14       1.13                 202.0   
3           1           0    1.06      0.35       2.29                 267.0   
4           1           0    0.81      0.51       1.31                 105.0   

   kills   k-t  k-ct  deaths  ...  tkmd-ct  kills assists survive trade perct  \
0   20.0  11.0   9.0    11.0  ...      6.0                               0.68   
1   16.0  12.0   4.0    13.0  ...      0.0                               0.84   
2   12.0   9.0   3.0    10.0  ...      1.0                               0.68   
3   18.0   8.0  10.0    13.0  ...      8.0                               0.79   
4    8.0   4.0   4.0    12.0  ...      2.0                               0.74   

   average damage per round  headshot percentage  first kill  fk-t  fk-ct  \
0                     160.0                 0.31         1.0   0.0    1.0   
1                     158.0                 0.41         0.0   0.0    0.0   
2                     148.0                 0.33         2.0   2.0    0.0   
3                     158.0                 0.27         7.0   3.0    4.0   
4                      64.0                 0.39         0.0   0.0    0.0   

   first death  fd-t  fd-ct  
0          1.0   0.0    1.0  
1          1.0   1.0    0.0  
2          1.0   1.0    0.0  
3          5.0   5.0    0.0  
4          1.0   0.0    1.0  

[5 rows x 27 columns]

### Displaying the first 5 rows of the dataset in order to set up basic predictions, I also made a copy of my workbook in excel with the exact data to make predictions easier to impute into the dataset

In [33]:
regressor.predict([[1.32,1.21,1.51,267,20,11,9,11,8,3,1,1,0,9,3,6,0.68,160,0.31,1,0,1,1,1,0]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([1.])

### Taken from the first row of the dataset to show the prediction of the model

In [34]:
regressor.predict([[1.34,1.51,1,314,20,14,6,11,7,4,1,1,0,9,7,2,.61,214,.35,0,0,0,2,1,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([0.])

### Taken from the 32nd row of the dataset to show the prediction of the model

In [35]:
from sklearn.metrics import r2_score
print('R-Square',r2_score(y_test,y_pred))

R-Square -0.9748167863081605


### R-Square value of the model to determine whats going on with the model

# With a negative r-squared value, and the use of this model I wont be able to get any meaning from r^2. 
# It is possible not able to capture the non-linear relationship between the features and the target variable. So I will take a Mean Square error test instead to gather more info. 

In [36]:
from sklearn.metrics import mean_squared_error as msc
from math import sqrt
rms=sqrt(msc(y_test,y_pred))
print('The Mean Square error is', rms)

The Mean Square error is 0.6935630188191351


### The mean square error of the model to determine the "accuracy" of the model
### With the low score of .6936 and an r-square that is negative its possible this model could work but I will want to gather more info and test other models before finilizing anything.

# Chaning the model to predict kills overall instead of just purely a win or a lose. 

In [37]:
X = df[['rating', 'rating-t', 'rating-ct', 'Average combat score', 'deaths',
        'd-t', 'd-ct', 'assists', 'a-t', 'a-ct', 'total kills minus deaths', 'tkmd-t', 'tkmd-ct', 'kills assists survive trade perct', 'average damage per round', 'headshot percentage', 'first kill', 'fk-t', 'fk-ct', 'first death', 'fd-t', 'fd-ct']]

### Chaning X to specific columns I need to predict kills

In [38]:
y = df['kills']

### Setting my Y as new predictor in kills

In [39]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)
regressor=DecisionTreeRegressor(random_state=42)
regressor.fit(X_train,y_train)

DecisionTreeRegressor(random_state=42)

### Retraining data

In [40]:
y_pred=regressor.predict(X_test)

### Predict the model

In [41]:
regressor.predict([[1.32,1.21,1.51,267,11,8,3,1,1,0,9,3,6,0.68,160,0.31,1,0,1,1,0,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([20.])

### Testing Model

In [42]:
regressor.predict([[1.34,1.51,1,314,11,7,4,1,1,0,9,7,2,.61,214,.35,0,0,0,2,1,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([20.])

### Testing Model

In [43]:
regressor.predict([[0.64,0.38,0.79,129,14,6,8,3,0,3,-5,-4,-1,0.68,87,0.27,1,0,1,0,0,0]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([9.])

### Testing Model

In [44]:
print('R-Square',r2_score(y_test,y_pred))

R-Square 0.9982068765623329


### New r^2 giving me almost a perfect score (Which isnt what i want normally a false positive)

In [45]:

rms=sqrt(msc(y_test,y_pred))
print('The Mean Square error is', rms)

The Mean Square error is 0.23459836628604097


### Super low MSE which tells me something is just wrong with this model because its too "perfect" 

In [50]:
regressor.score(X_test, y_test)
#accuracy score 

0.9982068765623329

# Switch to predicting a more apparent stats (Combat Score in the game means a lot more than most statistics)

In [51]:
X = df[['rating', 'rating-t', 'rating-ct', "kills", 'deaths',
        'd-t', 'd-ct', 'assists', 'a-t', 'a-ct', 'total kills minus deaths', 'tkmd-t', 'tkmd-ct', 'kills assists survive trade perct', 'average damage per round', 'headshot percentage', 'first kill', 'fk-t', 'fk-ct', 'first death', 'fd-t', 'fd-ct']]

In [52]:
y = df['Average combat score']

In [53]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)
regressor=DecisionTreeRegressor(random_state=42)
regressor.fit(X_train,y_train)

DecisionTreeRegressor(random_state=42)

In [54]:
y_pred=regressor.predict(X_test)

In [55]:
regressor.predict([[1.32,1.21,1.51,10,11,8,3,1,1,0,9,3,6,0.68,160,0.31,1,0,1,1,0,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([293.])

In [56]:
regressor.predict([[1.34,1.51,1,15,11,7,4,1,1,0,9,7,2,.61,214,.35,0,0,0,2,1,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([321.])

In [57]:
regressor.predict([[1.34,1.51,1,35,11,7,4,1,1,0,9,7,2,.61,214,.35,0,0,0,2,1,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([314.])

In [58]:
rms=sqrt(msc(y_test,y_pred))
print('The Mean Square error is', rms)

The Mean Square error is 19.237061105563814


In [59]:
regressor.score(X_test, y_test)
#accuracy score 

0.9033767775826569

# Starting prediction for a less impactful stat

In [60]:
X = df[['Average combat score', 'rating-t', 'rating-ct', "kills", 'deaths',
        'd-t', 'd-ct', 'assists', 'a-t', 'a-ct', 'total kills minus deaths', 'tkmd-t', 'tkmd-ct', 'kills assists survive trade perct', 'average damage per round', 'headshot percentage', 'first kill', 'fk-t', 'fk-ct', 'first death', 'fd-t', 'fd-ct']]

In [61]:
y = df['rating']

In [62]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)
regressor=DecisionTreeRegressor(random_state=42)
regressor.fit(X_train,y_train)

DecisionTreeRegressor(random_state=42)

In [63]:
y_pred=regressor.predict(X_test)

In [64]:
regressor.predict([[312,1.21,1.51,10,11,8,3,1,1,0,9,3,6,0.68,160,0.31,1,0,1,1,0,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([1.32])

In [65]:
regressor.predict([[255,1.51,1,35,11,7,4,1,1,0,9,7,2,.61,214,.35,0,0,0,2,1,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([1.28])

In [66]:
rms=sqrt(msc(y_test,y_pred))
print('The Mean Square error is', rms)

The Mean Square error is 0.08376269033222385


In [67]:
regressor.score(X_test, y_test)
#accuracy score 

0.994015793010674

# It appears that statistics that have more depth in their meaning like Average combat score works better with the decision tree model becuase although it is still numerical data inputed those stats can also be somewhat thought of as categorical data due to the difference in meanings between a low number and a high number. 

### Simply for testing I am going to use a linear regression model to see what type of accuracy I would get out putted for the higher vs lower impact stats to gauge and see if my theory is correct when using decision tree models for more categorical driven data. 

In [79]:
X = df[['rating', 'rating-t', 'rating-ct', "kills", 'deaths',
        'd-t', 'd-ct', 'assists', 'a-t', 'a-ct', 'total kills minus deaths', 'tkmd-t', 'tkmd-ct', 'kills assists survive trade perct', 'average damage per round', 'headshot percentage', 'first kill', 'fk-t', 'fk-ct', 'first death', 'fd-t', 'fd-ct']]

In [80]:
y = df['Average combat score']

In [81]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [82]:
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()


In [83]:
regressor.fit(X_train, y_train)


LinearRegression()

In [73]:
y_pred=regressor.predict(X_test)

In [84]:
regressor.predict([[1.32,1.21,1.51,10,11,8,3,1,1,0,9,3,6,0.68,160,0.31,1,0,1,1,0,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([234.07929383])

In [85]:
regressor.predict([[1.34,1.51,1,15,11,7,4,1,1,0,9,7,2,.61,214,.35,0,0,0,2,1,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([304.16901083])

In [86]:
print('R-Square',r2_score(y_test,y_pred))

R-Square -10.591920644858481


In [87]:
rms=sqrt(msc(y_test,y_pred))

print('The Mean Square error is', rms)

The Mean Square error is 210.70539982672148


# So for combat score we had a accuracy of 90% when using decision tree and then when inputed into the LR as a nuetral comparison we got an accuracy of -10% with a MSE of 210. So as my hypothesis was corret in that Decision tree is a good fit for this model when it comes to predicting average combat score statistics.

### As a final test I am going to compare the decision tree predictions for something like a lower impact statistic of assists for an exmpale. 

In [88]:
X = df[['rating', 'rating-t', 'rating-ct', "kills", 'deaths',
        'd-t', 'd-ct', 'a-t', 'a-ct', 'total kills minus deaths', 'tkmd-t', 'tkmd-ct', 'kills assists survive trade perct', 'average damage per round', 'headshot percentage', 'first kill', 'fk-t', 'fk-ct', 'first death', 'fd-t', 'fd-ct']]

In [89]:
y = df['assists']

In [92]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [93]:
regressor = LinearRegression()

In [94]:
regressor.fit(X_train, y_train)

LinearRegression()

In [95]:
y_pred=regressor.predict(X_test)

In [96]:
regressor.predict([[1.32,1.21,1.51,10,11,8,3,1,0,9,3,6,0.68,160,0.31,1,0,1,1,0,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([1.07722869])

In [97]:
regressor.predict([[1.34,1.51,1,15,11,7,4,1,0,9,7,2,.61,214,.35,0,0,0,2,1,1]])

C:\Users\Connor\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([1.04512354])

In [98]:
print('R-Square',r2_score(y_test,y_pred))

R-Square 0.9954641354650733


In [99]:
rms=sqrt(msc(y_test,y_pred))

print('The Mean Square error is', rms)

The Mean Square error is 0.2342436645692238


# Final Thoughts

### Overall this comparison showed that the main model with Decision Tree works well enough. Decision tree appears to excel with a combination of categorical data and numerical data which is somewhat similar to what I have here. Although the data is numerical stats like Average Combat Score has meaning behinding it at different levels which can be argued to be categorical. With the comparison to linear regression the accuracy was considerably lower in LR when dealing with Average combat score and when dealing with the less impactful stats like kills and assists linear regression was not drasitcally better by anymeans. Linear regression was not used in this case as a potential model for this data only as a nuteral comparison to the decision tree model. 

# Real World Application for the Model

### Based on the results of the decision tree model I would use this as a major stepping stone for my fantasy league predictions app. Using the previous data of the pros you could plot previous average combat scores of pros giving viewers a general overview of what they have been playing like in their previous games. Then you could use the "Average Combat Score Decision Tree model" to predict and plot future games combat scores to give a good representation of what their next couple games could look like. That would allow me to complete my objective of giving views and players a good understand and basis of how the pros play to be able to use that data in future fantasy leagues.